### ISSUE TITLE
bridge and masqurade networking doesn't work, but slirp does

### ISSUE THREAD

**Issue by dren-dk** [kubevirt/kubevirt#6820]
/kind bug

**What happened**:

Starting a VM with an interface in either bridge or masquerade mode gives a broken network where the default gateway is 169.254.1.1 and the VM is unable to ping it or connect to anything on the network.

If I use slirp, then networking seems to work correctly.

The VM has worked perfectly a long time on kubevirt v0.35.0 / kubernetes v1.16.3 / ubuntu 18.04 with network mode bridge.

**What you expected to happen**:

I would expect both bridge and masquerade to work.

**Anything else we need to know?**:

**Environment**:
- KubeVirt version (use `virtctl version`):
Client Version: version.Info{GitVersion:"v0.47.1", GitCommit:"c34de42a48f5564f4fd2c21b6cbda7b96664c65b", GitTreeState:"clean", BuildDate:"2021-11-11T16:01:45Z", GoVersion:"go1.16.6", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{GitVersion:"v0.47.1-dirty", GitCommit:"c34de42a48f5564f4fd2c21b6cbda7b96664c65b", GitTreeState:"dirty", BuildDate:"2021-11-11T16:20:45Z", GoVersion:"go1.16.6", Compiler:"gc", Platform:"linux/amd64"}

- Kubernetes version (use `kubectl version`):

Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.5", GitCommit:"6b1d87acf3c8253c123756b9e61dac642678305f", GitTreeState:"clean", BuildDate:"2021-03-18T01:10:43Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.5", GitCommit:"aea7bbadd2fc0cd689de94a54e5b7b758869d691", GitTreeState:"clean", BuildDate:"2021-09-15T21:04:16Z", GoVersion:"go1.16.8", Compiler:"gc", Platform:"linux/amd64"}

- VM or VMI specifications:

This VM doesn't work, the only difference between this and one that works is that bridge is replaced with slirp

`
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  annotations:
    kubevirt.io/latest-observed-api-version: v1
    kubevirt.io/storage-observed-api-version: v1alpha3
  generation: 1
  labels:
    app.kubernetes.io/instance: golden-virtual-machines
  managedFields:
  - apiVersion: kubevirt.io/v1alpha3
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .: {}
          f:kubectl.kubernetes.io/last-applied-configuration: {}
        f:labels:
          .: {}
          f:app.kubernetes.io/instance: {}
      f:spec:
        .: {}
        f:running: {}
        f:template:
          .: {}
          f:metadata:
            .: {}
            f:annotations: {}
            f:labels: {}
          f:spec:
            .: {}
            f:domain:
              .: {}
              f:cpu:
                .: {}
                f:cores: {}
              f:devices:
                .: {}
                f:disks: {}
                f:interfaces: {}
              f:machine:
                .: {}
                f:type: {}
              f:memory:
                .: {}
                f:guest: {}
              f:resources:
                .: {}
                f:requests:
                  .: {}
                  f:memory: {}
            f:networks: {}
            f:nodeSelector:
              .: {}
              f:kubernetes.io/hostname: {}
            f:volumes: {}
    manager: argocd-application-controller
    operation: Update
    time: "2021-11-19T12:01:10Z"
  - apiVersion: kubevirt.io/v1alpha3
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          f:kubevirt.io/latest-observed-api-version: {}
          f:kubevirt.io/storage-observed-api-version: {}
      f:status:
        .: {}
        f:conditions: {}
        f:created: {}
        f:printableStatus: {}
        f:ready: {}
        f:volumeSnapshotStatuses: {}
    manager: Go-http-client
    operation: Update
    time: "2021-11-19T12:01:34Z"
  name: win10
  namespace: golden-virtual-machines
  resourceVersion: "21294670"
  uid: 093633db-59dc-4c1a-8055-21e7b8538999
spec:
  running: true
  template:
    metadata:
      annotations:
        debugLogs: "true"
      creationTimestamp: null
      labels:
        debugLogs: "true"
        vmName: win10
    spec:
      domain:
        cpu:
          cores: 2
        devices:
          disks:
          - disk:
              bus: virtio
            name: boot
          - disk:
              bus: virtio
            name: service
          - disk:
              bus: virtio
            name: tools
          - cdrom:
              bus: sata
            name: secret-volume
          interfaces:
          - bridge: {}
            macAddress: 02:05:11:b0:42:42
            name: default
            ports:
            - port: 22
            - port: 80
            - port: 443
            - port: 3389
            - port: 3400
        machine:
          type: q35
        memory:
          guest: 8Gi
        resources:
          requests:
            memory: 8Gi
      networks:
      - name: default
        pod: {}
      nodeSelector:
        kubernetes.io/hostname: glengrant
      volumes:
      - name: boot
        persistentVolumeClaim:
          claimName: win10
      - containerDisk:
          image: registry.stibo.dk/rm/public/kubevirt/build-agent-service:11e66bd66fe017989f4f46bacbfe924fdf2ac6c8
        name: service
      - containerDisk:
          image: registry.stibo.dk/containers/public/windows-vm-tools-container:2a4b5912eec127f9ee910fd4c532bfa7e30feb5b
        name: tools
      - name: secret-volume
        secret:
          secretName: service-mode
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2021-11-19T12:01:31Z"
    status: "True"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: null
    message: 'cannot migrate VMI: PVC win10 is not shared, live migration requires
      that all PVCs must be shared (using ReadWriteMany access mode)'
    reason: DisksNotLiveMigratable
    status: "False"
    type: LiveMigratable
  created: true
  printableStatus: Running
  ready: true
  volumeSnapshotStatuses:
  - enabled: false
    name: boot
    reason: 'No VolumeSnapshotClass: Volume snapshots are not configured for this
      StorageClass [cephfs] [boot]'
  - enabled: false
    name: service
    reason: Snapshot is not supported for this volumeSource type [service]
  - enabled: false
    name: tools
    reason: Snapshot is not supported for this volumeSource type [tools]
  - enabled: false
    name: secret-volume
    reason: Snapshot is not supported for this volumeSource type [secret-volume]
`

- Cloud provider or hardware configuration:

Bare metal RKE v1.3.1 cluster with metallb in BGP mode.

The hosts are connected with ipoib (Infiniband).

- OS (e.g. from /etc/os-release):

Ubuntu 20.04.3 LTS

- Kernel (e.g. `uname -a`):

Linux glengrant 5.4.0-89-generic #100-Ubuntu SMP Fri Sep 24 14:50:10 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

**Comment by dren-dk** [comment]
I have since upgraded kubevirt on the old cluster and bridge works ok there, so that means that the working setup is:

* Kubevirt v0.35.0 and v0.47.1
* Kubernetes v1.16.3
* Docker 20.10.7
* Ubuntu 18.04
* Linux 4.15.0-128-generic
* Metallb in layer 2 mode

The setup where bridge and masqurade is broken is:
 
* Kubevirt v0.47.1
* Kubernetes v1.21.5
* Docker 20.10.10
* Ubuntu 20.04
* Linux 5.4.0-89-generic
* Metallb in BGP mode

**Comment by rmohr** [comment]
@dren-dk I see that you updated ubuntu there. Could it be related to app-amor? Just a first guess.

**Comment by rmohr** [comment]
/assign @maiqueb 

thanks for following up here.

**Comment by dren-dk** [comment]
The new cluster does have apparmor running, but so did the old one:
> sudo cat /sys/kernel/security/apparmor/profiles
docker-default (enforce)
/usr/sbin/ntpd (enforce)
/usr/sbin/tcpdump (enforce)
/{,usr/}sbin/dhclient (enforce)
/usr/lib/connman/scripts/dhclient-script (enforce)
/usr/lib/NetworkManager/nm-dhcp-helper (enforce)
/usr/lib/NetworkManager/nm-dhcp-client.action (enforce)
lsb_release (enforce)
man_groff (enforce)
man_filter (enforce)
/usr/bin/man (enforce)
nvidia_modprobe (enforce)
nvidia_modprobe//kmod (enforce)
/usr/sbin/haveged (enforce)

I've seen some references to the "/usr/libexec/qemu-kvm PUx," magic line to put in /etc/apparmor.d/usr.sbin.libvirtd, like:
https://github.com/kubevirt/kubevirt/issues/4303 but that issue seems to indicate that qemu cannot start at all.

**Comment by dren-dk** [comment]
I looked at the dmesg to see if apparmor metioned anything about it messing things up, but this is all that happens:

[Thu Nov 25 12:31:53 2021] eth0-nic: renamed from eth0
[Thu Nov 25 12:31:53 2021] k6t-eth0: port 1(eth0-nic) entered blocking state
[Thu Nov 25 12:31:53 2021] k6t-eth0: port 1(eth0-nic) entered disabled state
[Thu Nov 25 12:31:53 2021] device eth0-nic entered promiscuous mode
[Thu Nov 25 12:31:53 2021] k6t-eth0: port 2(tap0) entered blocking state
[Thu Nov 25 12:31:53 2021] k6t-eth0: port 2(tap0) entered disabled state
[Thu Nov 25 12:31:53 2021] device tap0 entered promiscuous mode
[Thu Nov 25 12:31:53 2021] k6t-eth0: port 2(tap0) entered blocking state
[Thu Nov 25 12:31:53 2021] k6t-eth0: port 2(tap0) entered forwarding state
[Thu Nov 25 12:31:53 2021] k6t-eth0: port 1(eth0-nic) entered blocking state
[Thu Nov 25 12:31:53 2021] k6t-eth0: port 1(eth0-nic) entered forwarding state

**Comment by dren-dk** [comment]
We uninstall snap and that left a dangling references in /etc/apparmor.d/usr.lib.snapd.snap-confine.real which caused
apparmor to not start, removing it and restarting apparmor didn't do anything, though.

I noticed that /etc/apparmor.d/usr.sbin.libvirtd didn't exist, so I lifted one from another system with libvirt-daemon-system installed, but restarting apparmor didn't do anything.

**Comment by sushiMix** [comment]
Hello I noticed a similar issue (for masquerade) but was not able to clearly troubleshoot it with default delivery.

I noticed that :
the ip route is not as expected
`
default via 169.254.1.1 dev eth0 
10.0.2.0/24 dev k6t-eth0 proto kernel scope link src 10.0.2.1 
169.254.1.1 dev eth0 scope link
`
the ip a command returns
`
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
3: eth0@if26: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1450 qdisc noqueue state UP group default 
    link/ether e6:2a:d2:c0:94:f2 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 10.244.61.60/32 brd 10.244.61.60 scope global eth0
       valid_lft forever preferred_lft forever
4: k6t-eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1450 qdisc noqueue state UP group default 
    link/ether 02:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff
    inet 10.0.2.1/24 brd 10.0.2.255 scope global k6t-eth0
       valid_lft forever preferred_lft forever
5: tap0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1450 qdisc pfifo_fast master k6t-eth0 state UP group default qlen 1000
    link/ether 7e:c4:78:ec:94:8f brd ff:ff:ff:ff:ff:ff
`

and as the ip route does not seem right I checked (nftables/iptables rules)
`nft list ruleset` is empty

 and 
`
bash-4.4# whoami
root
bash-4.4# iptables -S
iptables v1.8.4 (nf_tables): Could not fetch rule set generation id: Permission denied (you must be root)
`
Which means that even the root account can't access to potables/nftables (I'm not sure this is an expected behavior).

I can't understand how the network could work.

I also noticed that on virt-handler, 
calling the vm itself 10.0.2.2 is responding
calling the vm through the pod ip is not responding.


I hope It could help solving this issue (I'm testing Kubevirt 0.51.0)

**Comment by maiqueb** [comment]
> Hello I noticed a similar issue (for masquerade) but was not able to clearly troubleshoot it with default delivery.
> 
> I noticed that : the ip route is not as expected `default via 169.254.1.1 dev eth0 10.0.2.0/24 dev k6t-eth0 proto kernel scope link src 10.0.2.1 169.254.1.1 dev eth0 scope link` the ip a command returns `1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000 link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00 inet 127.0.0.1/8 scope host lo valid_lft forever preferred_lft forever` `3: eth0@if26: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1450 qdisc noqueue state UP group default link/ether e6:2a:d2:c0:94:f2 brd ff:ff:ff:ff:ff:ff link-netnsid 0 inet 10.244.61.60/32 brd 10.244.61.60 scope global eth0 valid_lft forever preferred_lft forever` `4: k6t-eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1450 qdisc noqueue state UP group default link/ether 02:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff inet 10.0.2.1/24 brd 10.0.2.255 scope global k6t-eth0 valid_lft forever preferred_lft forever` `5: tap0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1450 qdisc pfifo_fast master k6t-eth0 state UP group default qlen 1000 link/ether 7e:c4:78:ec:94:8f brd ff:ff:ff:ff:ff:ff`
> 
> and as the ip route does not seem right I checked (nftables/iptables rules) `nft list ruleset` is empty
> 
> and `bash-4.4# whoami` `root` `bash-4.4# iptables -S` `iptables v1.8.4 (nf_tables): Could not fetch rule set generation id: Permission denied (you must be root)` Which means that even the root account can't access to potables/nftables (I'm not sure this is an expected behavior).

I assume this is happening on the `virt-launcher` pod; that is expected: virt-launchers miss the `NET_ADMIN` capability, and as such, cannot perform these operations. That is a security feature.

If you want to access the netfilter rules, you have to access them via the corresponding `virt-handler` - which **does** have the required permissions. You first need to find the PID of the `virt-launcher` you want to debug, and then tap into its network namespace via `nsenter`:

```bash
nsenter -t <virt launcher PID> -n iptables -S
```

I am unsure if netfilter rules were configured via iptables / nftables. Try both :) 

> 
> I can't understand how the network could work.

You can find detailed information about how masquerade works in https://github.com/kubevirt/kubevirt/blob/main/docs/devel/networking.md#masquerade-binding-mechanism

It is probably a bit out-dated, and does not account for `istio` service mesh, but it is still accurate when it comes to the overall traffic flow. Especially interesting is the following sentence regarding the default route:
```
All outbound traffic from the VM can reach the outside world via the in-pod bridge. 
The packet will be routed via the default route - default via 169.254.1.1 dev eth0 - and
before leaving the interface, its source address will be masqueraded to the IP address
of the pod, via the masquerade target.
```

> 
> I also noticed that on virt-handler, calling the vm itself 10.0.2.2 is responding calling the vm through the pod ip is not responding.
> 
> I hope It could help solving this issue (I'm testing Kubevirt 0.51.0)

**Comment by sushiMix** [comment]
Thank you for your comment (and yes I was looking in virt-launcher pod), I'll check the documentation you pointed.


But I noticed this in the referenced documentation which is contrary to your comment (out of dated documentation ?).
<img width="1031" alt="image" src="https://user-images.githubusercontent.com/53741704/159680189-fb197287-b885-44e8-a7c7-9843c67df0fe.png">

**Comment by maiqueb** [comment]
> Thank you for your comment (and yes I was looking in virt-launcher pod), I'll check the documentation you pointed.
> 
> But I noticed this in the referenced documentation which is contrary to your comment (out of dated documentation ?). <img alt="image" width="1031" src="https://user-images.githubusercontent.com/53741704/159680189-fb197287-b885-44e8-a7c7-9843c67df0fe.png">

That is outdated - and as such, is no longer true. For more details check out the implementation of the pod templates, in [this link](https://github.com/kubevirt/kubevirt/blob/09450770a8f5dcc2905938b7c8f164d880346f68/pkg/virt-controller/services/template.go#L1809).

I've pushed PR #7431 to update the docs with the correct capability list.

**Comment by sushiMix** [comment]
Thank you for the document update

I tried to run command in the namespace but the nft or iptables always return an empty set whereas I expect forwarding rule :|

The inbound communication seems blocked whereas the outgoing is ok.

I'm using a simple sample 
```
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  name: testvm
  namespace: default
spec:
  running: true
  template:
    spec:
      domain:
        devices:
          disks:
            - disk:
                bus: virtio
              name: containerdisk
            - disk:
                bus: virtio
              name: cloudinitdisk
          interfaces:
            - masquerade: {}
              name: default
        machine:
          type: q35
        resources:
          requests:
            memory: 64M
      networks:
        - name: default
          pod: {}
      volumes:
        - containerDisk:
            image: quay.io/kubevirt/cirros-container-disk-demo
          name: containerdisk
        - cloudInitNoCloud:
            userDataBase64: SGkuXG4=
          name: cloudinitdisk
```

**Comment by maiqueb** [comment]
> `nsenter -t 15907 nft list ruleset` or `nsenter -t 15907 iptables -L` returns no specific rules (no Kubevirt ones)

The commands you've listed about are missing `-n`, which indicates you want to enter the **network** namespace. Without that, it won't work @sushiMix . 

Can you confirm you're actually tapping into the network namespace of the pod ? 

Would you paste the output of:
```bash
nsenter -t <launcher-pid> -n ip a
nsenter -t <launcher-pid> -n ip r
nsenter -t <launcher-pid> -n iptables -L
nsenter -t <launcher-pid> -n nft list chains ip
```

**Comment by sushiMix** [comment]
Hello sorry for the missing "-n" I don't usually manipulate the namespaces.

```
bash-4.4# nsenter -t 22646 -n ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
3: eth0@if27: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1450 qdisc noqueue state UP group default 
    link/ether 92:6d:ae:e9:6d:14 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 10.244.61.1/32 brd 10.244.61.1 scope global eth0
       valid_lft forever preferred_lft forever
4: k6t-eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1450 qdisc noqueue state UP group default 
    link/ether 02:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff
    inet 10.0.2.1/24 brd 10.0.2.255 scope global k6t-eth0
       valid_lft forever preferred_lft forever
5: tap0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1450 qdisc pfifo_fast master k6t-eth0 state UP group default qlen 1000
    link/ether 9a:da:e4:cf:fd:cf brd ff:ff:ff:ff:ff:ff
```

```
default via 169.254.1.1 dev eth0 
10.0.2.0/24 dev k6t-eth0 proto kernel scope link src 10.0.2.1 
169.254.1.1 dev eth0 scope link
```

```
nsenter -t 22646 -n iptables -L  
Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination 
```

```
bash-4.4# nsenter -t 22646 -n nft list chains ip
table ip nat {
        chain prerouting {
                type nat hook prerouting priority dstnat; policy accept;
        }
        chain input {
                type nat hook input priority 100; policy accept;
        }
        chain output {
                type nat hook output priority -100; policy accept;
        }
        chain postrouting {
                type nat hook postrouting priority srcnat; policy accept;
        }
        chain KUBEVIRT_PREINBOUND {
        }
        chain KUBEVIRT_POSTINBOUND {
        }
}
table ip filter {
        chain INPUT {
                type filter hook input priority filter; policy accept;
        }
        chain FORWARD {
                type filter hook forward priority filter; policy accept;
        }
        chain OUTPUT {
                type filter hook output priority filter; policy accept;
        }
}
```
but for nft I prefer
```
nsenter -t 22646 -n nft list ruleset   
table ip nat {
        chain prerouting {
                type nat hook prerouting priority dstnat; policy accept;
                iifname "eth0" counter packets 0 bytes 0 jump KUBEVIRT_PREINBOUND
        }

        chain input {
                type nat hook input priority 100; policy accept;
        }

        chain output {
                type nat hook output priority -100; policy accept;
                ip daddr { 127.0.0.1 } counter packets 0 bytes 0 dnat to 10.0.2.2
        }

        chain postrouting {
                type nat hook postrouting priority srcnat; policy accept;
                ip saddr 10.0.2.2 counter packets 3 bytes 202 masquerade
                oifname "k6t-eth0" counter packets 1 bytes 404 jump KUBEVIRT_POSTINBOUND
        }

        chain KUBEVIRT_PREINBOUND {
                counter packets 0 bytes 0 dnat to 10.0.2.2
        }

        chain KUBEVIRT_POSTINBOUND {
                ip saddr { 127.0.0.1 } counter packets 0 bytes 0 snat to 10.0.2.1
        }
}
table ip filter {
        chain INPUT {
                type filter hook input priority filter; policy accept;
        }

        chain FORWARD {
                type filter hook forward priority filter; policy accept;
        }

        chain OUTPUT {
                type filter hook output priority filter; policy accept;
        }
}
```
thanks for helping me.

**Comment by maiqueb** [comment]
I think the rules look correct ... @rhrazdil could you take a quick look at the nftables rules listed in the bottom of [this comment](https://github.com/kubevirt/kubevirt/issues/6820#issuecomment-1077317984) ? 

I find extremely strange that the pre routing rules was never hit:
`iifname "eth0" counter packets 0 bytes 0 jump KUBEVIRT_PREINBOUND`

That means traffic never hit the pod. @sushiMix, when you say "The inbound communication seems blocked whereas the outgoing is ok." from where are you trying to access the VM ?...

**Comment by rhrazdil** [comment]
Indeed, according to the routing rules, the inbound traffic never reached the pod.

For outbound traffic, we can see several connections in `postrouting` chain:
`ip saddr 10.0.2.2 counter packets 3 bytes 202 masquerade`.

**Comment by sushiMix** [comment]
> If you try it from within the virt-launcher pod, you should use the VM's static IP - which for masquerade is 10.0.2.2.

I agree this test was fully functional but tried after the "public" pod ip.

For the counters I don't remember If I did all the tests before outputting the notable rules.

**Comment by maiqueb** [comment]
> > If you try it from within the virt-launcher pod, you should use the VM's static IP - which for masquerade is 10.0.2.2.
> 
> I agree this test was fully functional but tried after the "public" pod ip.
> 
> For the counters I don't remember If I did all the tests before outputting the nft rules.

Please double check this :) It's important.

One more question: could it be that when you're reaching a random pod from a node, said random pod is running on that node ? And when you're trying to reach a VM from a node the VM is scheduled on a different node ? 

If so, would you try to reach out the VM from the node where it is running on ? 

My point is that according to the [Kubernetes network model](https://kubernetes.io/docs/concepts/services-networking/#the-kubernetes-network-model) : 
> agents on a node (e.g. system daemons, kubelet) can communicate with all pods on that node

I interpret this as not all CNIs will be able to reach pods on different nodes. Which reminds me to ask you: what is the default cluster network CNI plugin you are using ?

**Comment by sushiMix** [comment]
Hello,

I'm using the calico CNI and tried from the host running the VM as I'm testing on a single node.

For the counters I can restart the vm and perform specific test (I don't know how to reset anonymous counters)

**Comment by maiqueb** [comment]
> Hello,
> 
> I'm using the calico CNI and tried from the host running the VM as I'm testing on a single node.

That destroys my half-baked theory ... Thanks for the reply.

**Comment by sushiMix** [comment]
I finally found my issue (between chair and keyboard :|).

I do my test in a vm (VirtualBox) which runs the kubernetes node. I forgot that my VM had two network interfaces and one was using NAT. I noticed that one of the ip of my VirtualBox VM collides with the IP of the VM (which obviously breaks the routing mechanism). 
I fixed it by setting
```
pod:
  vmNetworkCIDR: 10.0.3.0/24
```

**Comment by dren-dk** [comment]
I have finally had a moment to grow angry enough about this problem to dig into it, here's what I have found:

* I installed a ubuntu server VM with slirp and checked that it works.
* I changed the VM to use bridging and checked that it failed as usual.
* I upgraded from kubevirt v0.48.0 to v0.58.0, just be sure that I would be chasing any fixed bugs.


Any regular (non-kubevirt) pod has the pod address set like this, with a subnet mask of /32:

```
3: eth0@if12: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1440 qdisc noqueue state UP group default 
    link/ether 9e:72:5a:bf:3e:4b brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 192.168.103.70/32 brd 192.168.103.70 scope global eth0
       valid_lft forever preferred_lft forever
```

Looking at the routes set up for the pod we get:

```
# ip route
default via 169.254.1.1 dev eth0 
169.254.1.1 dev eth0 scope link 
```

iow: There's a completely unrelated ip address that's explicitly reachable via the pod interface which is used as the default gw.


In the following I'm booting a plain ubuntu server VM, which is set up with DHCP that works as expected when using slirp, but it fails when using bridged networking.


I have been digging into the bridge configuration of the virt-launcher pod and I see that it does set up the bridge as expected:

```
brctl  show
bridge name     bridge id               STP enabled     interfaces
k6t-eth0                8000.0a9cfb35f8dc       no              eth0-nic
                                                                                   tap0
```

The default interface is down and the bridge (k6t-eth0) is up:

```
bash-4.4# ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
2: tunl0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN group default qlen 1000
    link/ipip 0.0.0.0 brd 0.0.0.0
4: eth0-nic@if29502: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1440 qdisc noqueue master k6t-eth0 state UP group default 
    link/ether 22:49:b6:dc:a5:3c brd ff:ff:ff:ff:ff:ff link-netnsid 0
5: eth0: <BROADCAST,NOARP> mtu 1500 qdisc noop state DOWN group default 
    link/ether 6e:77:d8:bb:c8:89 brd ff:ff:ff:ff:ff:ff
    inet 192.168.100.231/32 brd 192.168.100.231 scope global eth0
       valid_lft forever preferred_lft forever
6: k6t-eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1440 qdisc noqueue state UP group default 
    link/ether 0a:9c:fb:35:f8:dc brd ff:ff:ff:ff:ff:ff
    inet 169.254.75.10/32 scope global k6t-eth0
       valid_lft forever preferred_lft forever
7: tap0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1440 qdisc fq_codel master k6t-eth0 state UP group default qlen 1000
    link/ether 0a:9c:fb:35:f8:dc brd ff:ff:ff:ff:ff:ff
```

The VM is configured to use the tap0 interface according to both the command line and the libvirt domain xml:
```
                <interface type="ethernet">
                        <source></source>
                        <target dev="tap0" managed="no"></target>
                        <model type="virtio-non-transitional"></model>
                        <mac address="02:05:11:b0:42:42"></mac>
                        <mtu size="1440"></mtu>
                        <alias name="ua-default"></alias>
                        <rom enabled="no"></rom>
                </interface>
```

So it would seem the network has been configured perfectly and that can be confirmed by manually assigning the original pod address to the VMs interface and setting the routes:

```
# ip addr add 192.168.100.231/32 dev enp1s0
# ip link set enp1s0 up
# ip route add 196.254.1.1 dev enp1s0 scope link
# ip route add default via 169.254.1.1
```

With this the pod is able to reach out into the world and I can also reach into the pod via a regular LoadBalancer address.

Switching networking to slirp works as expected, so the dhcp client in the VM clearly works, but switching to bridged networking (or masqurade) causes the network configuration of the VM to fail.

It seems to me that the DHCP server is broken, as no IP or routes are assigned to the VM, unfortunately there's very little log output, even with verbosity turned up to 11, so this is just about all the DHCP related output of the virt-launcher log:

```
{"component":"virt-launcher","level":"info","msg":"The imported dhcpConfig: DHCPConfig: { Name: eth0, IPv4: 192.168.100.231/32 eth0, IPv6: \u003cnil\u003e, MAC: 02:05:11:b0:42:42, AdvertisingIPAddr: 169.254.75.10, MTU: 1440, Gateway: 169.254.1.1, IPAMDisabled: false}","pos":"podnic.go:260","timestamp":"2022-12-14T07:21:50.083040Z"}
{"component":"virt-launcher","level":"info","msg":"StartDHCP network Nic: DHCPConfig: { Name: eth0, IPv4: 192.168.100.231/32 eth0, IPv6: \u003cnil\u003e, MAC: 02:05:11:b0:42:42, AdvertisingIPAddr: 169.254.75.10, MTU: 1440, Gateway: 169.254.1.1, IPAMDisabled: false}","pos":"common.go:332","timestamp":"2022-12-14T07:21:50.083144Z"}
{"component":"virt-launcher","level":"info","msg":"Starting SingleClientDHCPServer","pos":"server.go:65","timestamp":"2022-12-14T07:21:50.083470Z"}
```

Rebooting the VM doesn't cause any log lines from the DHCP server as would be expected (with verbosity set to 11), but the virt-launcher is clearly listening for requests:

```
# netstat -lnpu
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
udp        0      0 0.0.0.0:67              0.0.0.0:*                           12/virt-launcher    
```

I tried running tcpdump in the VM and configured its network manager for verbose logging, but I never see any dhcp requests being sent from the VM.

At this point I reviewed the network configuration of the VM itself and it turned out that the network interface changed its name, so the VM was actually not configured to run DHCP, when I added the new network interface to the network manager configuration the VM started working perfectly.


The only reason I created the ubuntu based VM was to debug this situation, so I went back to the windows VMs that originally failed and after switching them to bridging they started working perfectly too.

Bottom line is that I think the bridging problem was fixed somewhere in the 0.48.0 to 0.58.0 time frame.

**Comment by xpivarc** [comment]
@dren-dk Nice debugging.
@EdDev Would you know of any changes that would fix this?

**Comment by AlonaKaplan** [comment]
> I do not see anything related, maybe @AlonaKaplan has a clue.

I would think that maybe this [PR](https://github.com/kubevirt/kubevirt/pull/5925). But it was fixed on v0.43.
The VM syslog is needed to be able to debug it. But since now it is working, I"m not sure it has to be debugged.
